### ISSUE TITLE
Parallel mode hangs when invoked from Rayon global thread pool

### ISSUE THREAD

**Issue by Pr0methean** [shssoichiro/oxipng#517]
My app needs to generate and optimize multiple images in parallel, so it uses the Rayon global thread pool. However, the time these images take varies unpredictably, so utilization at the end would be higher if I could also use Oxipng's parallel mode. Unfortunately, Oxipng deadlocks when I invoke oxipng from within a Rayon task in the global thread pool. Here's the app where I'm experiencing this: https://app.circleci.com/pipelines/github/Pr0methean/OcHd-RustBuild?branch=oxipng-parallel

**Comment by Pr0methean** [comment]
I've tried wrapping the `optimize_from_memory` call in `ThreadPool::install` with an explicitly-created thread pool, but that leads to starvation of the global pool and out-of-memory errors. A workaround would be if oxipng made the interface of `RowFilter` public, in which case I could create a separate task for each filter algorithm that should be tried.

**Comment by Pr0methean** [comment]
Master.

**Comment by andrews05** [comment]
Maybe this is relevant? https://github.com/rayon-rs/rayon/issues/592
Oxipng explicitly avoids using mutexes in this way, but perhaps your code does?

**Comment by Pr0methean** [comment]
That bug is unlikely to be relevant, since the PNG-writing task that invokes Oxipng calls into_inner() on its Mutex before running.

I'm not using the raw API but will look into it.

**Comment by andrews05** [comment]
As I'm not exactly an expert in Rayon, I'm afraid there's not much else I can do to help, sorry. Perhaps @kornelski may have some ideas?

**Comment by wait-what** [comment]
I am experiencing the same issue. https://github.com/mutant-remix/mrxbuilder/issues/1

Using this code as an example. My CPU has 12 threads. This works perfectly well with 11, but as soon as I uncomment 12, none of them finish, there is no CPU activity and the program just hangs.

```rs
use rayon::prelude::*;

use image::{codecs::png::PngEncoder, ImageEncoder, RgbaImage};
use oxipng::{optimize_from_memory, Options, Deflaters};

pub fn encode(rgba: &RgbaImage) -> Vec<u8> {
    let mut buffer = Vec::new();

    let encoder = PngEncoder::new(&mut buffer);
    encoder
        .write_image(&rgba, rgba.width(), rgba.height(), image::ColorType::Rgba8)
        .unwrap();

    let oxipng_options = Options {
        deflate: Deflaters::Libdeflater { compression: 12 },
        ..Default::default()
    };

    optimize_from_memory(&buffer, &oxipng_options).unwrap()
}


fn main() {
    let img = image::open("sample.png").unwrap().to_rgba8();

    vec![
        1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11,
        // 12,
    ].par_iter().for_each(|x| {
        println!("Encoding {x}");
        encode(&img);
        println!("Done {x}");
    });
}
```

**Comment by andrews05** [comment]
@wait-what Thanks for minimal test case, that's quite helpful!
It seems the problem is with the evaluator - trials are being scheduled in the background and then later it waits for them to finish but there are no threads executing them so it deadlocks.
I'm not really sure what the correct way to resolve this is, but here's one possible solution: https://github.com/andrews05/oxipng/commit/059e22ba203043bed0447155c9c205995d0dff4d
Would you and @Pr0methean mind checking this out to see if it works for you?

**Comment by wait-what** [comment]
Using `oxipng = { git = "https://github.com/andrews05/oxipng", branch = "parallel" }`. Using the same code, the issue seems to be fixed. Thank you! Hope it gets merged soon.

**Comment by kornelski** [comment]
I think the yield loop should be safe, because it's not busy-waiting, it's stealing work to run on the current thread. 

I'm not sure if it solves the problem 100%, because I don't know if it's guaranteed that rayon will find a job to execute at the moment of entering the `while` loop. If some timing issue made the evaluation job enqueue a bit later, I suspect the `while` loop may terminate, and then still get stuck waiting for the channel.

When `get_best_candidate` is called in a rayon context, it creates a situation where one rayon task is waiting for another rayon task to complete, and that is the equivalent of the mutex problem that rayon can't solve.

I see a couple of solutions to this:

1. Have an explicit separate rayon threadpool just for the `try_image` tasks, so that it's always guaranteed there's a thread available for them. 

2. In try_image have an explicit queue for images left to evaluate. Instead of spawning `rayon::task` for a specific image do this:
   1. Add an image to a queue (could be `Mutex<Vec<Arc<PngImage> & params>>`)
   2. Spawn rayon::task that removes an image from the queue and evaluates it

   And then in `get_best_candidate`, try to take images from the queue too, and eval them on the current thread too. This way if `rayon::task` never runs, `get_best_candidate` will finish the job. It's about the same as the `while rayon::yield` loop, but only for image tasks, and guaranteed to empty that particular queue.

**Comment by andrews05** [comment]
Thanks for the analysis, @kornelski!

> I'm not sure if it solves the problem 100%, because I don't know if it's guaranteed that rayon will find a job to execute at the moment of entering the `while` loop. If some timing issue made the evaluation job enqueue a bit later, I suspect the `while` loop may terminate, and then still get stuck waiting for the channel.

Are you sure this is possible though? I would have thought the job would be guaranteed to be queued before `rayon::spawn` returns. I mean, my personal preference would be to stick with the simplest solution unless we know for sure that it may not always work 😅

**Comment by andrews05** [comment]
@Pr0methean Have you had a chance to try out the branch above?

**Comment by TPS** [comment]
Is this work useful toward #165/#275, perchance?

**Comment by andrews05** [comment]
@TPS Not really, sorry. Although since #169 was opened we've added new filter types, so it will use up to 10 threads now at `-o 6`.

**Comment by Pr0methean** [comment]
@andrews05 It works; thanks!

**Comment by andrews05** [comment]
Great, I've opened a PR for it.

**Comment by andrews05** [comment]
@Pr0methean Was just thinking about the stack overflow you mentioned, are you able to get a stack trace for that?

**Comment by Pr0methean** [comment]
Yes.
[stack_trace.txt](https://github.com/shssoichiro/oxipng/files/11860800/stack_trace.txt)
The problem occurs because yield_local is effectively recursive.

**Comment by Pr0methean** [comment]
Opened https://github.com/rayon-rs/rayon/issues/1064.

**Comment by andrews05** [comment]
Oh I see, so yield_local is picking up a task from your own program (the next image to operate on), rather than an evaluation task, is that right?

**Comment by Pr0methean** [comment]
That's right. I've tried having the image task itself call yield_local() before starting, but this seems to be slower than without the Oxipng parallelization.

**Comment by andrews05** [comment]
I guess kornelski's solution would have worked better then. Whoops 😅
It is a lot more complex though, so hopefully rayon accepts your patch.

**Comment by Pr0methean** [comment]
Update: it looks like this only affects jobs running in a FIFO scope, but it's unfortunate that I have to choose between my two methods of increasing utilization (parallelizing oxipng vs starting the longest jobs first).

**Comment by andrews05** [comment]
I believe I have a simple solution. Can you try this out? https://github.com/andrews05/oxipng/commit/1e46122b0f7d7f5d26add02fc16dacf04261ec64

**Comment by Pr0methean** [comment]
It may take me a day or two to test this, because I'm also testing https://github.com/zopfli-rs/zopfli/pull/22 and (in preparation for making it a PR) https://github.com/Pr0methean/zopfli/commit/b2d014204af9f62c9f612944f9e4cfc6e64a4800.
